### PR TITLE
feat: sheafification on essentially small sites

### PR DESCRIPTION
**Pull request by dagurtomas**
This PR provides the necessary API to sheafify over large but essentially small categories. 

- Given an equivalence of categories and a Grothendieck topology on one of them, we transport the topology along the equivalence and prove that the sheaf categories are equivalent.
- We define a typeclass `HasSheafify` which says that presheaves on a site with values in some category can be sheafified, i.e. that the inclusion functor from sheaves to presheaves has a left exact left adjoint. We redefine `presheafToSheaf` as an arbitrary choice of such a left adjoint.
- We prove that the property `HasSheafify` transports over the equivalence of sheaf categories mentioned above, and same for the property `HasSheafCompose` 
- We also add instances for the existence of small limits on sheaf categories on essentially small sites.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Sites/Equivalence.lean [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+noncomputable
+def transportIsoSheafToPresheaf : (e.sheafCongr J).functor ⋙
+    sheafToPresheaf (e.locallyCoverDense J).inducedTopology A ⋙
+    e.op.congrLeft.inverse ≅ sheafToPresheaf J A :=
+  NatIso.ofComponents (fun F ↦ isoWhiskerRight e.op.unitIso.symm F.val)
+    (by intros; ext; simp [Equivalence.sheafCongr])
+
+/-- Transporting and sheafifying is left adjoint to taking the underlying presheaf. -/
+noncomputable
+def transportSheafificationAdjunction : transportAndSheafify J e ⊣ sheafToPresheaf J A :=
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Sites/Equivalence.lean:190]: These results are great, but is not it possible to consider introducing a type class saying that `sheafToPresheaf J A` has a left adjoint? and then redefine `presheafToSheaf` as a choice of left adjoint?

Comment (dagurtomas) [reply Mathlib/CategoryTheory/Sites/Equivalence.lean:190]: That's indeed a better philosophy. I'll try it out!

**Comment by dagurtomas** [comment]
Is this acceptable? It removes the following four lemmas:
In `CategoryTheory/Sites/Adjunction`:
- `CategoryTheory.Sheaf.adjunctionToTypes_unit_app_val`
- `CategoryTheory.Sheaf.adjunctionToTypes_counit_app_val`
In `CategoryTheory/Sites/CoverLifting`:
- `CategoryTheory.Functor.pushforwardContinuousSheafificationCompatibility_hom_app_val` and a helper lemma to prove it. 

They are never used in mathlib. The first two might be used in LTE, but I'm not sure. @adamtopaz added all of them originally.

If we want to keep these lemmas, we can, but then they would refer to the "explicit" sheafification functor which I renamed to `plusPlusSheaf`. The name `presheafToSheaf` now refers to an arbitrary choice of left adjoint to the inclusion

**Comment by erdOne** [comment]
Should `CategoryTheory.GrothendieckTopology.sheafify` be generalized as well?

**Comment by dagurtomas** [comment]
> Should `CategoryTheory.GrothendieckTopology.sheafify` be generalized as well?

How do you mean? That's just defined as the plus construction twice, and is the explicit construction used to prove that for a concrete category `D` with sufficiently nice properties, we have `HasSheafify J D`. I don't think there is a way to avoid this construction completely.

**Comment by erdOne** [comment]
If we decide to hide the double-plus construction as an implementation detail, we should probably define `CategoryTheory.GrothendieckTopology.sheafify` as `(presheafToSheaf J D).obj F`?

**Comment by dagurtomas** [comment]
I'll try that, and see where mathlib needs `sheafify` to actually mean the double-plus construction

**Comment by dagurtomas** [comment]
I did all the `sheafify` stuff in this new context. I opened #9012  which I'll hopefully finish tomorrow, where I just do the refactor of sheafification, changing as little as possible in the existing files.

**Comment by dagurtomas** [comment]
Replaced by #9523

**Pull request closed without merging**